### ISSUE TITLE
Infinite call to pbmclapply()

### ISSUE THREAD

**Issue by pat-s** [kvnkuang/pbmcapply#16]
HI @kvnkuang,

in the following example I get an infinite execution of `pbmclapply()`. Using `mclapply()` I get "at least" `NULL` returned. 

This problem only happens on macOS with a normal R startup. Using R in Vanilla mode solves it. 
Up to now its unclear what exactly causes this behaviour (http://stackoverflow.com/questions/44058387/r-mclapply-pblapply-vs-lapply-use-case). 
It also works fine on Linux. Anyway, to be able to deal with the `NULL` output and return an informative error message, it would be important that `pbmclapply()` does not run infinite. 

I would like to use your package in `parsperrorest()` as one parallel mode option.

`remotes::install_github("pat-s/sperrorest@performance")`

```r 
library(MASS)
library(sperrorest)
library(pbmcapply)

currentSample <- partition.cv(maipo, nfold = 4)
currentSample[[2]] <- partition.cv(maipo, nfold = 4)[[1]]
currentRes <- currentSample

lda.predfun <- function(object, newdata, fac = NULL) {
  library(nnet)
  majority <- function(x) {
    levels(x)[which.is.max(table(x))]
  }
  
  majority.filter <- function(x, fac) {
    for (lev in levels(fac)) {
      x[ fac == lev ] <- majority(x[ fac == lev ])
    }
    x
  }
  
  pred <- predict(object, newdata = newdata)$class
  if (!is.null(fac)) pred <- majority.filter(pred, newdata[,fac])
  return(pred)
}

data("maipo", package = "sperrorest")
predictors <- colnames(maipo)[5:ncol(maipo)]
fo <- as.formula(paste("croptype ~", paste(predictors, collapse = "+")))


# pbmclapply
runreps_res <- pbmclapply(cl = 2, currentSample, function(X) 
  runreps(currentSample = X, data = maipo,
          formula = fo, par.mode = 1, pred.fun = lda.predfun,
          do.try = FALSE, model.fun = lda,
          error.fold = TRUE, error.rep = TRUE, do.gc = 1,
          err.train = TRUE, importance = FALSE, currentRes = currentRes, 
          pred.args = list(fac = "field"), response = "croptype", par.cl = 2, 
          coords = c("x", "y"), progress = 1, pooled.obs.train = c(), 
          pooled.obs.test = c(), err.fun = err.default))

# mclapply
runreps_res <- mclapply(cl = 2, currentSample, function(X) 
  runreps(currentSample = X, data = maipo,
          formula = fo, par.mode = 1, pred.fun = lda.predfun,
          do.try = FALSE, model.fun = lda,
          error.fold = TRUE, error.rep = TRUE, do.gc = 1,
          err.train = TRUE, importance = FALSE, currentRes = currentRes, 
          pred.args = list(fac = "field"), response = "croptype", par.cl = 2, 
          coords = c("x", "y"), progress = 1, pooled.obs.train = c(), 
          pooled.obs.test = c(), err.fun = err.default))
```

**Comment by kvnkuang** [comment]
Hi pat-s,

Thanks for your feedback. The infinite running problem is usually caused by some internal error in `mcapply` that crashes the `futureCall()`.

I will need to add some error checking and handling routines so that when an internal error is triggered, the function would return or stop instead of running indefinitely.

Regards,
Kevin

**Comment by pat-s** [comment]
Great! Once this is fixed I can replace `mclapply()` by `pbmclapply()` in my package. Thanks for your work!

**Comment by kvnkuang** [comment]
Hi pat-s,

I pushed an update to the repo with the fix of error handling. Before submitting to CRAN, I'd like to make sure it solves your issue. 

Could you please install the development build and let me know whether it fixes your issue? Please using these commands to install dev build:
```
library(devtools)
install_github("kvnkuang/pbmcapply")
```

Thanks,
Kevin

**Comment by kvnkuang** [comment]
This issue has been addressed with 99fc925 - now online in version 1.2.2.

**Comment by pat-s** [comment]
works, thanks for the quick fix!
